src/objects.c

```diff
@@ -1323,57 +1423,100 @@ void accept_cancel_request(PgSocket *req)
+		needs_forwarding_to_peer = cf_peer_id != peer_id;
+		forwarded_from_peer = (req->cancel_key[7] & 1) == 0;
+		/*
+		 * Make sure the last bit is 1 when peering is enabled. All stored
+		 * cancelation keys have 1 for the last bit. A trailing 0 bit is only
+		 * used to communicate that the cancel request was forwarded.
+		 */
+		req->cancel_key[7] |= 1;
+	}
+
 
 	Assert(req->state == CL_LOGIN);
 
-	/* find real client this is for */
-	statlist_for_each(pitem, &pool_list) {
-		pool = container_of(pitem, PgPool, head);
-		statlist_for_each(citem, &pool->active_client_list) {
-			client = container_of(citem, PgSocket, head);
-			if (memcmp(client->cancel_key, req->cancel_key, 8) == 0) {
-				main_client = client;
-				goto found;
+	if (!needs_forwarding_to_peer) {
+		/* find real client this is for */
+		statlist_for_each(pitem, &pool_list) {
+			pool = container_of(pitem, PgPool, head);
+			statlist_for_each(citem, &pool->active_client_list) {
+				client = container_of(citem, PgSocket, head);
+				if (memcmp(client->cancel_key, req->cancel_key, 8) == 0) {
+					main_client = client;
+					goto found;
+				}
 			}
-		}
-		statlist_for_each(citem, &pool->waiting_client_list) {
-			client = container_of(citem, PgSocket, head);
-			if (memcmp(client->cancel_key, req->cancel_key, 8) == 0) {
-				main_client = client;
-				goto found;
+			statlist_for_each(citem, &pool->waiting_client_list) {
+				client = container_of(citem, PgSocket, head);
+				if (memcmp(client->cancel_key, req->cancel_key, 8) == 0) {
+					main_client = client;
+					goto found;
+				}
 			}
 		}
-	}
 found:
 
-	/* wrong key */
-	if (!main_client) {
+		/* wrong key */
+		if (!main_client) {
+			disconnect_client(req, false, "failed cancel request");
+			return;
+		}
+
+		/* not linked client, just drop it then */
+		if (!main_client->link) {
+			/* let administrative cancel be handled elsewhere */
+			if (main_client->pool->db->admin) {
+				disconnect_client(req, false, "cancel request for console client");
+				admin_handle_cancel(main_client);
+				return;
+			}
+
+			disconnect_client(req, false, "cancel request for idle client");
+
+			return;
+		}
+
+		/* remember server key */
+		server = main_client->link;
+		memcpy(req->cancel_key, server->cancel_key, 8);
+	} else if (forwarded_from_peer) {
 		disconnect_client(req, false, "failed cancel request");
 		return;
-	}
-
-	/* not linked client, just drop it then */
-	if (!main_client->link) {
-		/* let administrative cancel be handled elsewhere */
-		if (main_client->pool->db->admin) {
-			disconnect_client(req, false, "cancel request for console client");
-			admin_handle_cancel(main_client);
+	} else {
+		PgDatabase *peer = find_peer(peer_id);
+		if (!peer) {
+			disconnect_client(req, false, "could not find peer to forward request to");
 			return;
 		}
+		log_debug("forwarding cancellation request to peer %d", peer_id);
 
-		disconnect_client(req, false, "cancel request for idle client");
+		pool = get_peer_pool(peer);
```

Comment:
could use a comment, e.g.
```c
/*
 * When using peering (multiple pgbouncers behind the same load balancer), we may receive 
 * cancellation messages that were intended for a peer via the load balancer. We propagate the
 * cancellation via the peer's pool, instead of the server pool.
 */
```